design/KR/kenward.Rmd

```diff
@@ -0,0 +1,402 @@
+---
+title: how to implement Kenward Roger degree of freedom?
+author: "Liming Li"
+output: html_document
+editor_options:
+  chunk_output_type: console
+---
+
+# Objective
+
+The objective is to design how we can implement Kenward-Roger degree of freedom.
+
+# Current design of mmrm package
+
+In current mmrm package, we provided Satterthwaite degree of freedom, through `df_1d` and `df_md`.
+
+# Design of Kenward-Roger degree of freedom
+
+## update to current functions
+
+* `df_1d` and `df_md` renamed to `df_1d_sat` and `df_md_sat`, do not export them
+* add a new function `df_1d` and `df_md` to include degree of freedom selection
+* implement `df_1d_kr` and `df_md_kr` (take extra arguments to conduct improved Kenward-Roger degree of freedom)
+
+## Further updates to the naming
+* consider renaming the `df_1d` and `df_md` to something like `beta_test_1d` and `beta_test_md`?(the names need to be discussed)
+* add deprecation notes on `df_1d` and `df_md`
+
+## pseudo code
+
+```{r, eval=FALSE}
+# df_1d
+formula <- FEV1 ~ ARMCD + ad(AVISIT | USUBJID)
+result <- mmrm(formula, fev_data, reml = FALSE)
```

Comment:
I think we will already need to have an argument here to signal `mmrm()` and inside C++ code that we also want the adjusted variance-covariance matrix of the fixed effects to be returned.

Reply:
I wonder if we should include these P, Q, R, W in the C++ part, as the `df_1d` part should be separated with the `mmrm` part. I think it will be complex for `df_1d` with KR to really rely on some of the calculations in C++ side of the model fitting while `df_1d` with SAT does not; We can, somehow use C++ to do the calculations instead of R side for performance considerations. The adjusted variance-covariance should not be implemented while fitting the object, as you may change the `df_1d` methods after you get the raw fit. Instead of fitting a new model with new adjusted variance covariance matrix, having them separately should be fine

Reply:
Ok separate is also fine, but I think we need to do the derivative calculations with automatic differentiation in C++ because for the inverse Sigma it won't be feasible manually. Plus we need the speed

Reply:
you are right the inverse can be slow; we can report some kind of sigma inverse in the C++ side for each subject(because subjects can have different visits); for other part, like derivatives, I think we have closed form why we need the automatic differentiation? we can complete it in C++ side (but not with AD)

Reply:
I don't think there is a closed form for the partial derivatives of Sigma^-1, so I think we need AD there.
Also AD will be less work for future additions of covariance structures. In terms of coding this is a hard part, but I think it will be worth the effort. Might need require separate experimentation with TMB's AD mechanism to figure this out

Reply:
```{r}
inverse_derivative <- function(sigma_inv, dsigma) {
  ret <- lapply(dsigma, function(x) {
    -sigma_inv %*% x %*% sigma_inv
  })
  return(ret)
}
```

if sigma is invertible, we have  partial( sigma * sigma_inverse)/partial(theta) = partial(identity matrix) / partial(theta)

and partial( sigma * sigma_inverse)/partial(theta) =  partial(sigma) / partial(theta)  * sigma_inverse + sigma * partial(sigma_inverse) / partial(theta)

so partial(sigma_inverse) / partial(theta) = - sigma_inverse * partial(sigma) / partial(theta) * sigma_inverse

partial(sigma) / partial(theta) here we have closed form and as long as we get the inverse of sigma everything is fine

Reply:
ah cool! I did not see that. Very good. Still I wonder if for the partial derivatives we can employ AD to save writing code for each structure. Would be worth trying

Reply:
there is auto jacobian and hessian in https://kaskr.github.io/adcomp/namespaceautodiff.html#aece87159d1d7bc9099eff6b751143db4 

it seems that we can flat the covariance matrix and the take the jacobian of it (and loop its values to take hessian which is used by R matrice)